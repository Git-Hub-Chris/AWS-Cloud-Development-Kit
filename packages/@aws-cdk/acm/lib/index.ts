export * from './certificate';
export * from './certificate-ref';

// The L1 library for AWS::CertificateManager:
export * from '../cfn/certificatemanager';
