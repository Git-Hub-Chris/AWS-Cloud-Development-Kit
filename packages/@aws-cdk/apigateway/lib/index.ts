// The L1 Library for AWS::ApiGateway:
export * from '../cfn/apigateway';
