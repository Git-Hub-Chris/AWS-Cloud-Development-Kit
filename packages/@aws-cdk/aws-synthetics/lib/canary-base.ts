import * as cdk from '@aws-cdk/core';

export interface ICanary extends cdk.IResource {
  // Should have all the properties and method signatures associated with a canary
}

/**
 * Base of a canary
 */
export abstract class CanaryBase extends cdk.Resource implements ICanary {
}