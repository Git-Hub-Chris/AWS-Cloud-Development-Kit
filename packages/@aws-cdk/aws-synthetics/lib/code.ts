import * as s3 from '@aws-cdk/aws-s3';
import { Construct } from '@aws-cdk/core';

export abstract class Code {

  /**
   * @returns `InlineCode` with inline code.
   * @param code The actual handler code (limited to 4KiB)
   */
  public static fromInline(code: string): InlineCode {
    return new InlineCode(code);
  }
  /**
   * TODO implement `fromAsset()`
   */
  // public static fromAsset(path: string, options?: s3_assets.AssetOptions): AssetCode {
  //   return new AssetCode(path, options);
  // }

  /**
   * @returns `S3Code` associated with the specified S3 object.
   * @param bucket The S3 bucket
   * @param key The object key
   * @param objectVersion Optional S3 object version
   */
  public static fromBucket(bucket: s3.IBucket, key: string, objectVersion?: string): S3Code {
    return new S3Code(bucket, key, objectVersion);
  }

  /**
   * Called when the canary is initialized to allow this object to bind
   * to the stack, add resources and have fun.
   *
   * @param scope The binding scope. Don't be smart about trying to down-cast or
   * assume it's initialized. You may just use it as a construct scope.
   */
  public abstract bind(scope: Construct): CodeConfig;
}

export interface CodeConfig {
  /**
   * The location of the code in S3 (mutually exclusive with `inlineCode`).
   */
  readonly s3Location?: s3.Location;

  /**
   * Inline code (mutually exclusive with `s3Location`).
   */
  readonly inlineCode?: string;
}

/**
 * Canary code from an inline string (limited to 4KiB).
 */
export class InlineCode extends Code {
  public readonly isInline = true;

  constructor(private code: string) {
    super();

    if (code.length === 0) {
      throw new Error('Canary inline code cannot be empty');
    }

    if (code.length > 4096) {
      throw new Error('Canary source is too large, must be <= 4096 but is ' + code.length);
    }
  }

  public bind(_scope: Construct): CodeConfig {
    return {
      inlineCode: this.code,
    };
  }
}

/**
 * Canary code from an S3 archive.
 */
export class S3Code extends Code {
  public readonly isInline = false;
  private bucketName: string;

  constructor(bucket: s3.IBucket, private key: string, private objectVersion?: string) {
    super();

    if (!bucket.bucketName) {
      throw new Error('bucketName is undefined for the provided bucket');
    }

    this.bucketName = bucket.bucketName;
  }

  public bind(_scope: Construct): CodeConfig {
    return {
      s3Location: {
        bucketName: this.bucketName,
        objectKey: this.key,
        objectVersion: this.objectVersion,
      },
    };
  }
}