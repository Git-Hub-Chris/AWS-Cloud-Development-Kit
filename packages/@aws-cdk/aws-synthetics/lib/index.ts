export * from './canary';
export * from './code';
export * from './schedule';
export * from './test-prop';

// AWS::Synthetics CloudFormation Resources:
export * from './synthetics.generated';
