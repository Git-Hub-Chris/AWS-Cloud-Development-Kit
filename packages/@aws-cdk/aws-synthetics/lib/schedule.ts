import { Duration } from '@aws-cdk/core';

//TODO: remove Abstract
/**
 * Schedule for canary runs
 */
export abstract class Schedule {

  /**
   * Construct a schedule that runs the canary once.
   */
  public static once(): Schedule {
    return new LiteralSchedule('rate(0 minutes)');
  }

  /**
   * Construct a schedule from a literal schedule expression. The expression must be in a `rate(number units)` format
   *
   * @param expression The expression to use.
   */
  public static expression(expression: string): Schedule {
    return new LiteralSchedule(expression);
  }

  /**
   * Construct a schedule from an interval. Allowed values: 0 (for a single run) or between 1 and 60 minutes.
   */
  public static rate(duration: Duration): Schedule {
    const minutes = duration.toMinutes();
    if (minutes === 0) {
      return Schedule.once();
    }
    if (minutes > 60) {
      throw new Error('Schedule duration must be between 1 and 60 minutes');
    }
    if (minutes === 1) {
      return new LiteralSchedule('rate(1 minute)');
    }
    return new LiteralSchedule(`rate(${minutes} minutes)`);
  }

  /**
   * Retrieve the expression for this schedule
   */
  public abstract readonly expressionString: string;
}

class LiteralSchedule extends Schedule {
  constructor(public readonly expressionString: string){
    super();
  }
}