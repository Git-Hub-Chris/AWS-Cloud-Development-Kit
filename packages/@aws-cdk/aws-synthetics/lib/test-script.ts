import { Code } from './code';
import * as s3 from '@aws-cdk/aws-s3';
import { Construct } from '@aws-cdk/core';

export class Test {
  /**
   * @returns `CustomTest` associated with the specified Code object
   * @param code The script you want the canary to run
   * @param handler The handler of the code
   */
  public static custom(scope: Construct, options: CustomOptions): Test {
    const codeConfig = options.code.bind(scope);
    return new Test({
      handler: options.handler,
      inlineCode: codeConfig.inlineCode,
      s3Location: codeConfig.s3Location,
    });
  }

  /**
   * TODO
   */
  //public static heartBeat(url: string){}

  /**
   * TODO
   */
  //public static apiEndpoint(url: string, options: apiOptions){}

  /**
   * TODO
   */
  //public static brokenLink(url: string, options: linkOptions){}

  private constructor(public readonly config: TestConfig){}
}

export interface TestConfig {
  /**
   * The location of the code in S3 (mutually exclusive with `inlineCode`).
   */
  readonly s3Location?: s3.Location;

  /**
   * Inline code (mutually exclusive with `s3Location`).
   */
  readonly inlineCode?: string;

  /**
   * The handler of the code
   */
  readonly handler: string;
}

// export class CustomTest extends Test {
//   constructor(public readonly code: Code, public readonly handler: string){
//     super();

//   }


// }

export interface CustomOptions {
  code: Code,
  handler: string,
}