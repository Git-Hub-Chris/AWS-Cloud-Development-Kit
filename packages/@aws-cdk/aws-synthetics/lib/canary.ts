import { Metric, MetricOptions } from '@aws-cdk/aws-cloudwatch';
import * as iam from '@aws-cdk/aws-iam';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import { Schedule } from './schedule';
import { CfnCanary } from './synthetics.generated';

/**
 * Runtime options for a canary
 */
export class Runtime {
  /**
   * `syn-1.0` includes the following:
   *
   * Synthetics library 1.0
   *
   * Synthetics handler code 1.0
   *
   * Lambda runtime Node.js 10.x
   *
   * Puppeteer-core version 1.14.0
   *
   * The Chromium version that matches Puppeteer-core 1.14.0
   */
  public static readonly SYN_1_0 = new Runtime('syn-1.0');

  private constructor(
    /**
     * The name of the runtime version
     */
    public readonly name: string){}
}

/**
 * Options for specifying the s3 location that stores the data of each canary run
 */
export interface ArtifactsBucketOptions {
  /**
   * The s3 location that stores the data of each run.
   */
  readonly bucket: s3.IBucket;

  /**
   * The key prefix of an S3 object. For example, if a key is specified, the resulting s3 location looks like this:
   * @example
   * s3://bucket/key
   *
   * If not specified, the s3 location looks like this:
   *
   * @example
   * s3://onlybucket
   */
  readonly prefix?: string;
}

/**
 * Properties for a canary
 */
export interface CanaryProps {
  /**
   * The s3 location that stores the data of the canary runs.
   *
   * @default - A new s3 bucket will be created.
   */
  readonly artifactsBucket?: ArtifactsBucketOptions;

  /**
   * Canary execution role.
   *
   * This is the role that will be assumed by the canary upon execution.
   * It controls the permissions that the canary will have. The Role must
   * be assumable by the AWS Lambda service principal.
   *
   * If not supplied, a role will be created with all the required permissions.
   * If you provide a Role, you must add the required permissions.
   *
   * @see required permissions: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-synthetics-canary.html#cfn-synthetics-canary-executionrolearn
   *
   * @default - A unique role will be generated for this canary.
   * You can add permissions to roles by calling 'addToRolePolicy'.
   */
  readonly role?: iam.IRole;

  /**
   * Specify the runtime version to use for the canary. Currently, the only valid value is `Runtime.SYN_1.0`.
   *
   * @default Runtime.SYN_1_0
   */
  readonly runtime?: Runtime;

  /**
   * How long the canary will be in a 'RUNNING' state. For example, if you set `timeToLive` to be 1 hour and `schedule` to be `rate(10 minutes)`,
   * your canary will run at 10 minute intervals for an hour, for a total of 6 times.
   *
   * @default - no limit
   */
  readonly timeToLive?: cdk.Duration;

  /**
   * Specify the schedule for how often the canary runs. For example, if you set `schedule` to `rate(10 minutes)`, then the canary will run every 10 minutes.
   * You can set the schedule with `Schedule.rate(Duration)` (recommended) or you can specify an expression using `Schedule.expression()`.
   * @default 'rate(5 minutes)'
   */
  readonly schedule?: Schedule;

  /**
   * Whether or not the canary should start after creation.
   *
   * @default true
   */
  readonly startAfterCreation?: boolean;

  /**
   * How many days should successful runs be retained.
   *
   * @default Duration.days(31)
   */
  readonly successRetentionPeriod?: cdk.Duration;

  /**
   * How many days should failed runs be retained.
   *
   * @default Duration.days(31)
   */
  readonly failureRetentionPeriod?: cdk.Duration;

  /**
   * The name of the canary. Be sure to give it a descriptive name that distinguishes it from
   * other canaries in your account.
   *
   * Do not include secrets or proprietary information in your canary name. The canary name
   * makes up part of the canary ARN, which is included in outbound calls over the internet.
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/servicelens_canaries_security.html
   *
   * @default - A unique name will be generated from the construct ID
   */
  readonly canaryName?: string;

  /**
   * Specify the endpoint that you want the canary code to hit. Alternatively, you can specify
   * your own canary script to run.
   *
   * ðŸš§ TODO: implement
   */
  //readonly test: Test;

}

/**
 * Define a new Canary
 */
export class Canary extends cdk.Resource {
  /**
   * Execution role associated with this Canary.
   */
  public readonly role: iam.IRole;

  /**
   * The canary ID
   * @attribute
   */
  public readonly canaryId: string;

  /**
   * The state of the canary. For example, RUNNING.
   * @attribute
   */
  public readonly canaryState: string;

  /**
   * The canary Name
   * @attribute
   */
  public readonly canaryName: string;

  /**
   * Bucket where data from each canary run is stored
   */
  private readonly artifactsBucket: s3.IBucket;

  public constructor(scope: cdk.Construct, id: string, props: CanaryProps) {
    if (props.canaryName && !cdk.Token.isUnresolved(props.canaryName)) {
      validateName(props.canaryName);
    }

    super(scope, id, {
      physicalName: props.canaryName || cdk.Lazy.stringValue({
        produce: () => this.generateUniqueName(),
      }),
    });

    this.artifactsBucket = props.artifactsBucket?.bucket ?? new s3.Bucket(this, 'ServiceBucket', {
      encryption: s3.BucketEncryption.KMS_MANAGED,
    });

    this.role = props.role ?? this.createDefaultRole();

    //this.artifactsBucket.grantWrite(this.role);

    const resource: CfnCanary = new CfnCanary(this, 'Resource', {
      artifactS3Location: this.artifactsBucket.s3UrlForObject(props.artifactsBucket?.prefix),
      executionRoleArn: this.role.roleArn,
      startCanaryAfterCreation: props.startAfterCreation !== false,
      runtimeVersion: props.runtime?.name ?? Runtime.SYN_1_0.name,
      name: this.physicalName,
      schedule: this.createSchedule(props),
      failureRetentionPeriod: props.failureRetentionPeriod?.toDays(),
      successRetentionPeriod: props.successRetentionPeriod?.toDays(),
      // ðŸš§ TODO: implement
      code: {
        handler: 'index.handler',
        script: 'exports.handler = async () => {\nconsole.log(\'hello world\');\n};',
      },
    });

    this.canaryId = resource.attrId;
    this.canaryState = resource.attrState;
    this.canaryName = this.getResourceNameAttribute(resource.ref);
  }

  /**
   * @param metricName - the name of the metric
   * @param options - configuration options for the metric
   *
   * @returns a CloudWatch metric associated with the canary.
   * @default avg over 5 minutes
   */
  private metric(metricName: string, options?: MetricOptions): Metric {
    return new Metric({
      metricName,
      namespace: 'CloudWatchSynthetics',
      dimensions: { CanaryName: this.canaryName },
      statistic: 'avg',
      ...options,
    }).attachTo(this);
  }

  /**
   * Measure the Duration of a single canary run, in seconds.
   *
   * @param options - configuration options for the metric
   *
   * @default avg over 5 minutes
   */
  public metricDuration(options?: MetricOptions): Metric {
    return this.metric('Duration', options);
  }

  /**
   * Measure the percentage of successful canary runs.
   *
   * @param options - configuration options for the metric
   *
   * @default avg over 5 minutes
   */
  public metricSuccessPercent(options?: MetricOptions): Metric {
    return this.metric('SuccessPercent', options);
  }

  /**
   * Measure the number of failed canary runs over a given time period.
   *
   * @param options - configuration options for the metric
   *
   * @default avg over 5 minutes
   */
  public metricFailed(options?: MetricOptions): Metric {
    return this.metric('Failed', options);
  }

  /**
   * Returns a default role for the canary
   */
  private createDefaultRole(): iam.IRole {
    // Created role will need these policies to run the Canary.
    // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-synthetics-canary.html#cfn-synthetics-canary-executionrolearn
    const policy = new iam.PolicyDocument({
      statements: [
        new iam.PolicyStatement({
          resources: ['*'],
          actions: ['s3:ListAllMyBuckets'],
        }),
        new iam.PolicyStatement({
          resources: [this.artifactsBucket.arnForObjects('*')],
          actions: ['s3:PutObject', 's3:GetBucketLocation'],
        }),
        new iam.PolicyStatement({
          resources: ['*'],
          actions: ['cloudwatch:PutMetricData'],
          conditions: {StringEquals: {'cloudwatch:namespace': 'CloudWatchSynthetics'}},
        }),
        new iam.PolicyStatement({
          resources: ['arn:aws:logs:::*'],
          actions: ['logs:CreateLogStream', 'logs:CreateLogGroup', 'logs:PutLogEvents'],
        }),
      ],
    });

    return new iam.Role(this, 'ServiceRole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      inlinePolicies: {
        canaryPolicy: policy,
      },
    });
  }

  /**
   * Returns a canary schedule object
   */
  private createSchedule(props: CanaryProps): CfnCanary.ScheduleProperty {
    return {
      durationInSeconds: String(`${props.timeToLive?.toSeconds() ?? 0}`),
      expression: props.schedule?.expressionString ?? 'rate(5 minutes)',
    };
  }

  /**
   * Creates a unique name for the canary. The generated name is the physical ID of the canary.
   */
  private generateUniqueName(): string {
    const name = this.node.uniqueId.toLowerCase().replace(' ', '-');
    if (name.length <= 21){
      return name;
    } else {
      const front = name.substring(0,15);
      const hash = this.hashCode(name.substring(15));
      return front + String(hash).substring(0,6);
    }
  }

  /**
   * @see https://stackoverflow.com/questions/6122571/simple-non-secure-hash-function-for-javascript
   * @param s the string to be hashed
   */
  private hashCode(s: string): number {
    var hash = 0;
    for(let i = 0; i<s.length; i++){
      hash = ((hash << 5) - hash) + s.charCodeAt(i); //eslint-disable-line no-bitwise
      // Convert to 32bit integer
      hash = hash & hash; //eslint-disable-line no-bitwise
    }
    return hash;
  }
}

const regex: RegExp = new RegExp('^[0-9a-z_\-]+$');

/**
 * Verifies that the name fits the regex expression: ^[0-9a-z_\-]+$.
 *
 * @param name - the given name of the canary
 */
function validateName(name: string) {
  if (name.length > 21) {
    throw new Error('Canary name is too large, must be <= 21 characters, but is ' + name.length);
  }
  if (!regex.test(name)) {
    throw new Error(`Canary name must be lowercase, numbers, hyphens, or underscores (no spaces) (${name})`);
  }
}
