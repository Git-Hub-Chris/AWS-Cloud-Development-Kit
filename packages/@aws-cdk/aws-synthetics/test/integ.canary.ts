import * as cdk from '@aws-cdk/core';
import * as synthetics from '../lib';

/*
 * Stack verification steps:
 *
 * -- aws synthetics get-canary --name integ-canary has a state of 'RUNNING'
 */
const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-canary');

new synthetics.Canary(stack, 'mycanary-124', {
  test: synthetics.Test.custom(stack, {
    code: synthetics.Code.fromInline('exports.handler = async () => {\nconsole.log(\'hello world\');\n};'),
    handler: 'index.handler',
  }),
});

app.synth();