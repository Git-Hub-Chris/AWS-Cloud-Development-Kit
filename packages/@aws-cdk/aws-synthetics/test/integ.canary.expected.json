{
  "Resources": {
    "mycanary555ServiceBucket40CDD613": {
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "mycanary555ServiceRoleF8E57AF2": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:ListAllMyBuckets",
                  "Effect": "Allow",
                  "Resource": "*"
                },
                {
                  "Action": [
                    "s3:PutObject",
                    "s3:GetBucketLocation"
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "mycanary555ServiceBucket40CDD613",
                      "Arn"
                    ]
                  }
                },
                {
                  "Action": "cloudwatch:PutMetricData",
                  "Condition": {
                    "StringEquals": {
                      "cloudwatch:namespace": "CloudWatchSynthetics"
                    }
                  },
                  "Effect": "Allow",
                  "Resource": "*"
                },
                {
                  "Action": [
                    "logs:CreateLogStream",
                    "logs:CreateLogGroup",
                    "logs:PutLogEvents"
                  ],
                  "Effect": "Allow",
                  "Resource": "arn:aws:logs:::*"
                }
              ],
              "Version": "2012-10-17"
            },
            "PolicyName": "canaryPolicy"
          }
        ]
      }
    },
    "mycanary555DAA1D9EA": {
      "Type": "AWS::Synthetics::Canary",
      "Properties": {
        "ArtifactS3Location": {
          "Fn::Join": [
            "",
            [
              "s3://",
              {
                "Ref": "mycanary555ServiceBucket40CDD613"
              }
            ]
          ]
        },
        "Code": {
          "Handler": "index.handler",
          "Script": "exports.handler = async () => {\nconsole.log('hello world');\n};"
        },
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "mycanary555ServiceRoleF8E57AF2",
            "Arn"
          ]
        },
        "Name": "awscanarymycanary555",
        "RuntimeVersion": "syn-1.0",
        "Schedule": {
          "DurationInSeconds": "0",
          "Expression": "rate(5 minutes)"
        },
        "StartCanaryAfterCreation": true,
        "RunConfig": {
          "MemoryInMB": 960,
          "TimeoutInSeconds": 300
        }
      }
    }
  }
}